File: JDK11/opentraceability/src/main/java/opentraceability/OTLogger.java
package opentraceability;

import java.util.ArrayList;
import java.util.List;

public class OTLogger {
    private static final List<OnLogDelegate> listeners = new ArrayList<>();

    public static void addListener(OnLogDelegate listener) {
        listeners.add(listener);
    }

    public static void removeListener(OnLogDelegate listener) {
        listeners.remove(listener);
    }

    public static void error(Exception ex) {
        OTLog log = new OTLog();
        log.Level = LogLevel.Error;
        log.Exception = ex;
        notifyListeners(log);
    }

    public static void notifyListeners(OTLog log) {
        for (OnLogDelegate listener : listeners) {
            listener.invoke(log);
        }
    }
}

File: Java/OpenTraceability/src/main/opentraceability/otlogger.kt

enum class LogLevel(val value: Int) {
    Info(0),
    Warning(1),
    Error(2),
    Debug(3),
}


typealias OnLogDelegate = (log: OTLog) -> Unit


object OTLogger {
    private val listeners: MutableList<OnLogDelegate> = mutableListOf()

    fun addListener(listener: OnLogDelegate) {
        listeners.add(listener)
    }

    fun removeListener(listener: OnLogDelegate) {
        listeners.remove(listener)
    }

    fun error(ex: Exception) {
        val log = OTLog().apply {
            Level = LogLevel.Error
            Exception = ex
        }
        notifyListeners(log)
    }

    private fun notifyListeners(log: OTLog) {
        listeners.forEach { listener ->
            listener(log)
        }
    }
}
